File: backend/src/models/sos-event.ts
/**
 * CEAL Backend — SOS Event model & Zod validation.
 *
 * V2 privacy-first protocol:
 *   {id, bleUid (hex), flags, sequence, timestamp, status, relayHops,
 *    receiverLocation: {lat, lon}, rssi?, message?}
 *
 * The BLE packet no longer contains GPS — the receiver attaches its own
 * location when forwarding to the backend.
 */

import { z } from 'zod';

// ---------------------------------------------------------------------------
// Status enum
// ---------------------------------------------------------------------------

export const SOS_STATUSES = [
  'active',
  'relayed',
  'acknowledged',
  'resolved',
  'cancelled',
] as const;

export type SosStatus = (typeof SOS_STATUSES)[number];

// ---------------------------------------------------------------------------
// Zod schemas — used to validate incoming JSON payloads
// ---------------------------------------------------------------------------

/**
 * Validates the POST /sos/ingest body (V2 protocol).
 */
export const sosIngestSchema = z.object({
  id: z.string().min(1).max(128),
  /** Hex-encoded 6-byte BLE UID (12 hex chars) */
  bleUid: z.string().regex(/^[0-9a-fA-F]{12}$/),
  /** BLE advertisement flags (bit 0 = SOS active, bit 1 = medical) */
  flags: z.number().int().min(0).max(255),
  /** Packet sequence number (0-255, wrapping) */
  sequence: z.number().int().min(0).max(255),
  timestamp: z.string().datetime({ offset: true }).or(z.string().datetime()),
  status: z.enum(SOS_STATUSES).default('active'),
  relayHops: z.number().int().min(0).default(0),
  message: z.string().max(64).optional(),
  /** Receiver location (lat/lon of the relay device, NOT the victim) */
  receiverLocation: z.object({
    lat: z.number().min(-90).max(90),
    lon: z.number().min(-180).max(180),
    accuracy: z.number().optional(),
  }).optional(),
  /** RSSI of BLE signal at receiver (dBm, negative) */
  rssi: z.number().int().optional(),
});

// ---------------------------------------------------------------------------
// SOS Type helper — extract type label from flags byte bits 2-5
// ---------------------------------------------------------------------------

const SOS_TYPES = [
  'General SOS',
  'Fire Emergency',
  'Crime Alert',
  'Kidnap Alert',
  'Medical Emergency',
  'Natural Disaster',
] as const;

/**
 * Extract a human-readable SOS type label from the flags byte.
 * Bits 2-5 encode the type code (0-15). Unknown codes default to 'General SOS'.
 */
export function extractSosType(flags: number): string {
  const code = (flags >> 2) & 0x0F;
  return SOS_TYPES[code] ?? 'General SOS';
}

/**
 * Validates the POST /sos/acknowledge body.
 */
export const sosAckSchema = z.object({
  id: z.string().min(1).max(128),
});

// ---------------------------------------------------------------------------
// TypeScript types
// ---------------------------------------------------------------------------

export interface SosEvent {
  id: string;
  bleUid: string;
  flags: number;
  sequence: number;
  /** ISO 8601 string */
  timestamp: string;
  status: SosStatus;
  relayHops: number;
  message?: string;
  /** Receiver location */
  receiverLat?: number;
  receiverLon?: number;
  rssi?: number;
  /** Resolved user ID (if BLE UID matched a registered user) */
  userId?: string;
  /** Human-readable SOS type label extracted from flags */
  sosType?: string;
}

export type SosIngestPayload = z.infer<typeof sosIngestSchema>;
export type SosAckPayload = z.infer<typeof sosAckSchema>;

// ---------------------------------------------------------------------------
// User-related types
// ---------------------------------------------------------------------------

export interface User {
  id: string;
  name: string | null;
  phone: string;
  bleUid: Buffer;
  language: string | null;
  createdAt: string;
}

export interface EmergencyContact {
  id: string;
  userId: string;
  name: string | null;
  phone: string | null;
  priority: number;
}

export interface MedicalProfile {
  userId: string;
  bloodGroup: string | null;
  allergies: string | null;
  conditions: string | null;
}

export interface FullUserProfile {
  user: User;
  contacts: EmergencyContact[];
  medical: MedicalProfile | null;
}

File: mobile/lib/models/sos_event.dart
/// SOS Event model — core domain object relayed over BLE and backend.
library;

import 'dart:typed_data';

import 'package:aftermath/core/constants.dart';
import 'package:aftermath/models/sos_type.dart';

enum SosStatus { active, relayed, acknowledged, resolved, cancelled }

class ReceiverLocation {
  const ReceiverLocation({required this.lat, required this.lon, this.accuracy});

  final double lat;
  final double lon;
  final double? accuracy;

  Map<String, dynamic> toJson() => {
    'lat': lat,
    'lon': lon,
    if (accuracy != null) 'accuracy': accuracy,
  };

  factory ReceiverLocation.fromJson(Map<String, dynamic> json) {
    return ReceiverLocation(
      lat: (json['lat'] as num).toDouble(),
      lon: (json['lon'] as num).toDouble(),
      accuracy: json['accuracy'] != null
          ? (json['accuracy'] as num).toDouble()
          : null,
    );
  }
}

class SosEvent {
  SosEvent({
    required this.id,
    required this.bleUid,
    required this.flags,
    required this.sequence,
    required this.timestamp,
    this.status = SosStatus.active,
    this.relayHops = 0,
    this.receiverLocation,
    this.rssi,
    this.message,
  });

  final String id;
  final Uint8List bleUid;
  final int flags;
  final int sequence;
  final DateTime timestamp;
  SosStatus status;
  int relayHops;
  ReceiverLocation? receiverLocation;
  int? rssi;
  final String? message;

  String get bleUidHex =>
      bleUid.map((b) => b.toRadixString(16).padLeft(2, '0')).join();

  Duration get age => DateTime.now().toUtc().difference(timestamp);
  bool get isExpired => age > kPacketMaxAge;

  /// SOS type extracted from bits 2-5 of the flags byte.
  SosType get sosType => SosType.fromFlags(flags);

  Map<String, dynamic> toJson() => {
    'id': id,
    'bleUid': bleUidHex,
    'flags': flags,
    'sequence': sequence,
    'timestamp': timestamp.toUtc().toIso8601String(),
    'status': status.name,
    'relayHops': relayHops,
    if (receiverLocation != null)
      'receiverLocation': receiverLocation!.toJson(),
    if (rssi != null) 'rssi': rssi,
    if (message != null) 'message': message,
  };

  factory SosEvent.fromJson(Map<String, dynamic> json) {
    final hex = json['bleUid'] as String? ?? '';
    final uid = Uint8List(kBleUidSize);
    if (hex.length >= kBleUidSize * 2) {
      for (int i = 0; i < kBleUidSize; i++) {
        uid[i] = int.parse(hex.substring(i * 2, i * 2 + 2), radix: 16);
      }
    }

    return SosEvent(
      id: json['id'] as String,
      bleUid: uid,
      flags: json['flags'] as int? ?? 0,
      sequence: json['sequence'] as int? ?? 0,
      timestamp: DateTime.parse(json['timestamp'] as String).toUtc(),
      status: SosStatus.values.byName((json['status'] as String?) ?? 'active'),
      relayHops: json['relayHops'] as int? ?? 0,
      receiverLocation: json['receiverLocation'] is Map<String, dynamic>
          ? ReceiverLocation.fromJson(
              json['receiverLocation'] as Map<String, dynamic>,
            )
          : null,
      rssi: json['rssi'] as int?,
      message: json['message'] as String?,
    );
  }

  @override
  String toString() {
    return 'SosEvent($id, $status, uid=$bleUidHex, seq=$sequence, hops=$relayHops)';
  }
}
